vaadin-flow-components-shared-parent/vaadin-flow-components-base/src/main/java/com/vaadin/flow/component/shared/DisableOnClickController.java

```diff
@@ -0,0 +1,135 @@
+     * <p>
+     * The component must implement {@link HasEnabled} and
+     * {@link ClickNotifier}.
+     *
+     * @param component
+     *            the component to control, not {@code null}
+     * @throws IllegalArgumentException
+     *             if the component does not implement {@link HasEnabled} or
+     *             {@link ClickNotifier}
+     */
+    public DisableOnClickController(Component component) {
+        this.component = Objects.requireNonNull(component);
+        if (!(component instanceof HasEnabled)) {
+            throw new IllegalArgumentException(
+                    "The component has to implement HasEnabled");
+        }
+        if (!(component instanceof ClickNotifier)) {
+            throw new IllegalArgumentException(
+                    "The component has to implement ClickNotifier");
+        }
+        ((ClickNotifier<Component>) component)
+                .addClickListener(itemClickEvent -> {
+                    if (disableOnClick) {
+                        ((HasEnabled) component).setEnabled(false);
+                    }
+                });
+        component.addAttachListener(aAttachEvent -> {
+            if (isDisableOnClick()) {
+                initDisableOnClick();
+            }
+        });
+    }
+
+    /**
+     * Sets whether the component should be disabled when clicked.
+     * <p>
+     * When set to {@code true}, the component will be immediately disabled on
+     * the client-side when clicked, preventing further clicks until re-enabled
+     * from the server-side.
+     *
+     * @param disableOnClick
+     *            whether the component should be disabled when clicked
+     */
+    public void setDisableOnClick(boolean disableOnClick) {
+        this.disableOnClick = disableOnClick;
+        if (disableOnClick) {
+            component.getElement().setAttribute("disableOnClick", "true");
+            initDisableOnClick();
+        } else {
+            component.getElement().removeAttribute("disableOnClick");
+        }
+    }
+
+    /**
+     * Gets whether the component is set to be disabled when clicked.
+     *
+     * @return whether the component is set to be disabled on click
+     */
+    public boolean isDisableOnClick() {
+        return disableOnClick;
+    }
+
+    /**
+     * Forces the client-side component's {@code disabled} property to be
+     * updated immediately.
+     * <p>
+     * This method should be called from the component's
+     * {@link HasEnabled#setEnabled} method.
+     *
+     * @param enabled
+     *            value to set
+     */
+    public void onSetEnabled(boolean enabled) {
+        // If the component is then disabled and re-enabled during the same
+        // round trip, Flow will not detect any changes and the client side
+        // component would not be enabled again.
+        component.getElement().executeJs("this.disabled = $0", !enabled);
+    }
+
+    private void initDisableOnClick() {
```

Comment:
Could keep the JavaDoc from `Button` here to provide some context.

Reply:
Done.